src/lf_stack.ml

```diff
@@ -0,0 +1,62 @@
+(** Trieber's Lock Free stack *)
+
+type 'a node =
+  | Nil
+  | Next of 'a * 'a node Atomic.t
+
+type 'a t =
+  { head : 'a node Atomic.t ;
+    tail : 'a node Atomic.t }
+
+let create () =
+  let head = (Next (Obj.magic (), Atomic.make Nil)) in
+  { head = Atomic.make head ; tail = Atomic.make head }
+
+let is_empty q =
+  match Atomic.get q.head with
+  | Nil -> failwith "Stack.is_empty: impossible"
+  | Next (_,x) ->
+     ( match Atomic.get x with
+       | Nil -> true
+       | _ -> false )
+
+let push q v =
+
+  let rec loop b () =
+    let head = Atomic.get q.head in
+    let newnode = Next (v, Atomic.make head) in
+    if (Atomic.compare_and_set q.head head newnode) then ()
+    else (Backoff.once b; loop b ()) (* This can fail we need to loop *)
+  in 
+
+  let head = Atomic.get q.head in
+  let newnode = Next (v, Atomic.make head) in
+  if (Atomic.compare_and_set q.head head newnode) then ()
+  else (
+    let b = Backoff.create () in
+    Backoff.once b;
+    loop b ())
+
+let pop q =
+  let rec loop b () =
+    let s = Atomic.get q.head in
+    let newhead = match s with
+      | Nil -> failwith "Stack.pop: impossible"
+      | Next (_, x) -> Atomic.get x
+    in match newhead with
+       | Nil -> None
+       | Next (v, _) when Atomic.compare_and_set q.head s newhead -> Some v
```

Comment:
Afaict this has ABA problem. If we read `newhead`, and in the meantime:
1. Other thread pops it
2. Other thread insert a new node, which won the lottery and got the same reference.
This CAS will succeed despite looking at a different node, and function will return a wrong value (old one). 

In the original implementation/paper this is solved by including a counter in the CAS. 

While theoretically possible, in our case gc would have to run between points 1., 2. to make that scenario possible. I wonder if there is any way to guarantee that that does not happen, e.g. disable safepoints or anything.

Reply:
DSCheck: 

```

 let test () = 
  let queue = create () in 
  push queue 1; 
  push queue 2;
  push queue 1; 
  push queue 2;
  push queue 1; 
  push queue 2;
  push queue 1; 
  push queue 2;
  Atomic.spawn (fun () -> 
    clean_until queue (fun v -> v == 1));
  Atomic.spawn (fun () -> 
    clean_until queue (fun v -> v == 2));
  Atomic.final (fun () -> 
    Atomic.check (fun () -> 
      not (is_empty queue)))
```

Reply:
Ah no - impossible since we hold the reference to `s` the whole time. Marking as resolved.